### PR TITLE
IPC4: fix capture issue in CI test

### PR DESCRIPTION
**Pull request by RanderWang**
This issue is found by CI capture test with Linux ipc4 driver.
 For a pipeline like module A <-> mixin <-> mixout, mixin is
first bound to mixout then bound to module A in CI test. The
original code assume that moudle A is first bound to mixin
 since windows driver always bind module from data source to
data sink.

Also need to update direction in each module

### CODE REVIEW

**Comment by lgirdwood** [comment]
@tmleman @marcinszkudlinski

File: src/audio/mixer.c [hunk 1]
```diff
@@ -743,33 +744,43 @@ static int mixin_bind(struct comp_dev *dev, void *data)
 
 	/* mixin -> mixout */
 	if (dev->ipc_config.id == src_id) {
-		struct list_item *blist;
+		if (list_is_empty(&dev->bsource_list))
```
Comment (lgirdwood) [inline src/audio/mixer.c:747]: can we have comment - why are we checking for empty list here.

File: src/audio/mixer.c [hunk 2]
```diff
@@ -743,33 +744,43 @@ static int mixin_bind(struct comp_dev *dev, void *data)
...
 
-		if (!sink_buf) {
-			comp_err(dev, "mixin_bind: no sink buffer found");
-			return -EINVAL;
+		sink_buf = buffer_from_list(dev->bsink_list.next, struct comp_buffer,
+					    PPL_DIR_DOWNSTREAM);
+		sink = buffer_get_comp(sink_buf, PPL_DIR_DOWNSTREAM);
+	}
+
+	list_for_item(blist, &sink->bsource_list) {
```
Comment (lgirdwood) [inline src/audio/mixer.c:766]: please add comments on what we are doing here.

File: src/audio/mixer.c [hunk 3]
```diff
@@ -743,33 +744,43 @@ static int mixin_bind(struct comp_dev *dev, void *data)
...
-		source_buf = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
-		pipeline_disconnect(dev, source_buf, PPL_CONN_DIR_BUFFER_TO_COMP);
-		pipeline_connect(sink, source_buf, PPL_CONN_DIR_BUFFER_TO_COMP);
-		pipeline_connect(dev, sink_buf, PPL_CONN_DIR_BUFFER_TO_COMP);
+	if (!sink_buf) {
+		comp_err(dev, "mixin_bind: no sink buffer found");
+		return -EINVAL;
 	}
 
+	source_buf = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
```
Comment (lgirdwood) [inline src/audio/mixer.c:779]: and here too - pls comment.

**Review by lgirdwood — COMMENTED** [review]
We need to comment the IPC4 logic so it's easy to follow.

**Comment by sys-pt1s** [comment]
Can one of the admins verify this patch?

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
test this please

**Comment by lgirdwood** [comment]
@wszypelt good to merge ?

**Comment by wszypelt** [comment]
@lgirdwood Yes, good to merge

**Pull request merged**